CMakeLists.txt

```diff
@@ -1,298 +1,28 @@
-    ARCHIVE  DESTINATION ${CMAKE_INSTALL_LIBDIR}
-    LIBRARY  DESTINATION ${CMAKE_INSTALL_LIBDIR}
-    RUNTIME  DESTINATION ${CMAKE_INSTALL_BINDIR})  # This is for Windows
-#install(DIRECTORY include/ DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
-
-add_library(${PROJECT_NAME}::${LIB_NAME} ALIAS ${LIB_NAME})
-
-# This makes the project importable from the install directory
-# Put config file in per-project dir (name MUST match), can also
-# just go into 'cmake'.
-install(
-	EXPORT ${PROJECT_NAME}-targets
-	DESTINATION share/${PROJECT_NAME}/cmake
-	NAMESPACE ${PROJECT_NAME}::
-)
-
-# This makes the project importable from the build directory
-export(
-	TARGETS ${LIB_NAME}
-	FILE ${PROJECT_NAME}-targets.cmake
-	NAMESPACE ${PROJECT_NAME}::
-)
-
-
-# Create variable with the library location
-set(libnabo_library $<TARGET_FILE:${LIB_NAME}>)
-
-# Create variable for the local build tree
-get_property(libnabo_include_dirs DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR} PROPERTY INCLUDE_DIRECTORIES)
-# Configure & generate config file for local build tree
-configure_file(libnaboConfig.cmake.in
-	"${PROJECT_BINARY_DIR}/libnaboConfig.cmake.conf" @ONLY)
-file(GENERATE
-	OUTPUT "${PROJECT_BINARY_DIR}/libnaboConfig.cmake"
-	INPUT "${PROJECT_BINARY_DIR}/libnaboConfig.cmake.conf")
-
-# 2- installation build #
-
-# Change the library location for an install location
-set(libnabo_library ${CMAKE_INSTALL_PREFIX}/lib/$<TARGET_FILE_NAME:${LIB_NAME}>)
-
-# Change the include location for the case of an install location
-set(libnabo_include_dirs ${CMAKE_INSTALL_PREFIX}/include)
-
-# We put the generated file for installation in a different repository (i.e., ./CMakeFiles/)
-configure_file(libnaboConfig.cmake.in
-	"${PROJECT_BINARY_DIR}${CMAKE_FILES_DIRECTORY}/libnaboConfig.cmake.conf" @ONLY)
-file(GENERATE
-	OUTPUT "${PROJECT_BINARY_DIR}${CMAKE_FILES_DIRECTORY}/libnaboConfig.cmake"
-	INPUT "${PROJECT_BINARY_DIR}${CMAKE_FILES_DIRECTORY}/libnaboConfig.cmake.conf")
-
-# The same versioning file can be used for both cases
-configure_file(libnaboConfigVersion.cmake.in
-	"${PROJECT_BINARY_DIR}/libnaboConfigVersion.cmake" @ONLY)
-
-install(FILES
-	"${PROJECT_BINARY_DIR}${CMAKE_FILES_DIRECTORY}/libnaboConfig.cmake"
-	"${PROJECT_BINARY_DIR}/libnaboConfigVersion.cmake"
-	DESTINATION share/libnabo/cmake COMPONENT dev)
-
-
-#=============================================
-# Add uninstall target
-#=============================================
-if (NOT TARGET uninstall)
-  configure_file(
-	  "${CMAKE_CURRENT_SOURCE_DIR}/cmake_uninstall.cmake.in"
-	  "${CMAKE_CURRENT_BINARY_DIR}/cmake_uninstall.cmake"
-	  IMMEDIATE @ONLY)
 
-  add_custom_target(uninstall
-	  COMMAND ${CMAKE_COMMAND} -P ${CMAKE_CURRENT_BINARY_DIR}/cmake_uninstall.cmake)
+file(GLOB_RECURSE NABO_HEADERS "nabo/*.h")
+
+# Global compiler flags.
+add_compile_options(-Wall -Wextra -Wpedantic)
+
+# Build type options:
+#	* catkin.
+#	* cmake.
+#	* ament (future).
+set(BUILD_TYPE "cmake")
```

Comment:
I suggest also testing if `CATKIN_DEVEL_PREFIX` cmake var is set, and automatically choose catkin build if it is. I've just tested that this variable is set at the very beginning of the build even before `find_package(catkin)`. And, it is the only one that has this property. Tested both with `catkin` and `catkin tools`.

Reply:
> Shouldn't this rather check a CMake variable and fill in cmake only in case it's unset or empty?

Can you give an example?

> I suggest also testing if CATKIN_DEVEL_PREFIX cmake var is set, and automatically choose catkin build if it is. I've just tested that this variable is set at the very beginning of the build even before find_package(catkin). And, it is the only one that has this property. Tested both with catkin and catkin tools.

I thought about something similar, but I think it's better to just allow the user to be explicit about the build type. I also don't know if e.g. the ethz-asl CI is already packaging libnabo as a sort-of ROS package with pure cmake build.

Reply:
Changing the build type requires editing the CMakeLists.txt (if this isn't some black magic that would turn this local variable into cached option). So at least expose the build type as an option, the same way USE_OPEN_MP and others do. And, when introducing the option, please, give it a unique name to prevent collisions with other packages - e.g. NABO_BUILD_TYPE.

I would still be happier if libnabo detected it is built using catkin and autoselected catkin build in that case. But I agree that can break some older setups that expect the pure CMake package. These could continue building using catkin and set NABO_BUILD_TYPE=cmake explicitly. Without the autodetection, all users who want the catkin package, would need to explicitly set the build type. It is doable, but not convenient. But this is more a political decision, you have to properly weigh all pros and cons.